src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -145,12 +145,19 @@ static SemanticTokensHelpers()
             // can pass in a range if they wish.
             var textSpan = range is null ? root.FullSpan : ProtocolConversions.RangeToTextSpan(range, text);
 
+            // Razor uses isFinalized to determine whether to cache tokens. We should be able to remove it altogether once
+            // Roslyn implements workspace/semanticTokens/refresh: https://github.com/dotnet/roslyn/issues/60441
+            var isFinalized = false;
+            if (document.Project.TryGetCompilation(out _))
+            {
+                var workspaceStatusService = document.Project.Solution.Workspace.Services.GetRequiredService<IWorkspaceStatusService>();
+                isFinalized = await workspaceStatusService.IsFullyLoadedAsync(cancellationToken).ConfigureAwait(false);
+            }
+
             // If the full compilation is not yet available, we'll try getting a partial one. It may contain inaccurate
             // results but will speed up how quickly we can respond to the client's request.
-            var frozenDocument = document.WithFrozenPartialSemantics(cancellationToken);
-            var semanticModel = await frozenDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
-            var isFinalized = document.Project.TryGetCompilation(out var compilation) && compilation == semanticModel.Compilation;
-            document = frozenDocument;
+            document = document.WithFrozenPartialSemantics(cancellationToken);
+            options = options with { ForceFrozenPartialSemanticsForCrossProcessOperations = true };
```

Comment:
❔ Do we want to use frozen partial semantics with `isFinalized`?

Reply:
Hmm, I wouldn't think so? I believe we want to call `TryGetCompilation` on the complete document. The only other place the frozen document would be used with `isFinalized` is when retrieving the WorkspaceStatusService, and unless I'm mistaken it doesn't seem like frozen partial semantics would be useful for that.